File: fraiseql-go/fraiseql/observers.go
package fraiseql

// ObserverAction represents an action to execute when an observer is triggered
type ObserverAction map[string]interface{}

// RetryConfig represents retry configuration for observer actions
type RetryConfig struct {
	MaxAttempts      int    `json:"max_attempts"`
	BackoffStrategy  string `json:"backoff_strategy"`
	InitialDelayMs   int    `json:"initial_delay_ms"`
	MaxDelayMs       int    `json:"max_delay_ms"`
}

// DefaultRetryConfig provides sensible retry defaults
var DefaultRetryConfig = RetryConfig{
	MaxAttempts:     3,
	BackoffStrategy: "exponential",
	InitialDelayMs:  100,
	MaxDelayMs:      60000,
}

// ObserverDefinition represents an observer definition
type ObserverDefinition struct {
	Name      string           `json:"name"`
	Entity    string           `json:"entity"`
	Event     string           `json:"event"`
	Actions   []ObserverAction `json:"actions"`
	Condition string           `json:"condition,omitempty"`
	Retry     RetryConfig      `json:"retry"`
}

// ObserverBuilder provides a fluent interface for building observers
type ObserverBuilder struct {
	name      string
	entity    string
	event     string
	actions   []ObserverAction
	condition string
	retry     RetryConfig
}

// NewObserver creates a new observer builder
func NewObserver(name string) *ObserverBuilder {
	return &ObserverBuilder{
		name:    name,
		retry:   DefaultRetryConfig,
		actions: []ObserverAction{},
	}
}

// Entity sets the entity type to observe
func (ob *ObserverBuilder) Entity(entity string) *ObserverBuilder {
	ob.entity = entity
	return ob
}

// Event sets the event type (INSERT, UPDATE, DELETE)
func (ob *ObserverBuilder) Event(event string) *ObserverBuilder {
	ob.event = event
	return ob
}

// Action adds an action to execute
func (ob *ObserverBuilder) Action(action ObserverAction) *ObserverBuilder {
	ob.actions = append(ob.actions, action)
	return ob
}

// Actions sets multiple actions at once
func (ob *ObserverBuilder) Actions(actions ...ObserverAction) *ObserverBuilder {
	ob.actions = actions
	return ob
}

// Condition sets the condition expression
func (ob *ObserverBuilder) Condition(condition string) *ObserverBuilder {
	ob.condition = condition
	return ob
}

// Retry sets the retry configuration
func (ob *ObserverBuilder) Retry(retry RetryConfig) *ObserverBuilder {
	ob.retry = retry
	return ob
}

// Register registers the observer with the global schema registry
func (ob *ObserverBuilder) Register() {
	definition := ObserverDefinition{
		Name:      ob.name,
		Entity:    ob.entity,
		Event:     ob.event,
		Actions:   ob.actions,
		Condition: ob.condition,
		Retry:     ob.retry,
	}

	RegisterObserver(definition)
}

// Webhook creates a webhook action
func Webhook(url string, options ...map[string]interface{}) ObserverAction {
	action := ObserverAction{
		"type":    "webhook",
		"headers": map[string]string{"Content-Type": "application/json"},
	}

	if url != "" {
		action["url"] = url
	}

	// Apply options
	if len(options) > 0 {
		opts := options[0]
		if urlEnv, ok := opts["url_env"].(string); ok {
			action["url_env"] = urlEnv
			delete(action, "url") // Use url_env instead
		}
		if headers, ok := opts["headers"].(map[string]string); ok {
			action["headers"] = headers
		}
		if bodyTemplate, ok := opts["body_template"].(string); ok {
			action["body_template"] = bodyTemplate
		}
	}

	return action
}

// WebhookWithEnv creates a webhook action using an environment variable
func WebhookWithEnv(urlEnv string, options ...map[string]interface{}) ObserverAction {
	opts := map[string]interface{}{"url_env": urlEnv}
	if len(options) > 0 {
		for k, v := range options[0] {
			opts[k] = v
		}
	}
	return Webhook("", opts)
}

// Slack creates a Slack notification action
func Slack(channel string, message string, options ...map[string]interface{}) ObserverAction {
	action := ObserverAction{
		"type":             "slack",
		"channel":          channel,
		"message":          message,
		"webhook_url_env":  "SLACK_WEBHOOK_URL",
	}

	// Apply options
	if len(options) > 0 {
		opts := options[0]
		if webhookURL, ok := opts["webhook_url"].(string); ok {
			action["webhook_url"] = webhookURL
			delete(action, "webhook_url_env")
		}
		if webhookURLEnv, ok := opts["webhook_url_env"].(string); ok {
			action["webhook_url_env"] = webhookURLEnv
		}
	}

	return action
}

// EmailAction creates an email action
// Note: Named EmailAction to avoid conflict with Email scalar type
func EmailAction(to string, subject string, body string, options ...map[string]interface{}) ObserverAction {
	action := ObserverAction{
		"type":    "email",
		"to":      to,
		"subject": subject,
		"body":    body,
	}

	// Apply options
	if len(options) > 0 {
		opts := options[0]
		if fromEmail, ok := opts["from_email"].(string); ok {
			action["from"] = fromEmail
		}
	}

	return action
}

File: sdks/official/fraiseql-typescript/src/observers.ts
/**
 * Observer authoring API for FraiseQL event-driven workflows.
 *
 * Observers watch for database changes and trigger actions (webhooks, Slack, email).
 * NO runtime behavior - only used for schema compilation.
 *
 * @example
 * ```typescript
 * import { Observer, webhook, slack } from "fraiseql";
 *
 * class Observers {
 *   @Observer({ entity: "Order", event: "INSERT", actions: [webhook("https://...")] })
 *   onOrderCreated() {}
 * }
 * ```
 */

import { SchemaRegistry, ObserverAction, ObserverRetryConfig } from "./registry";

/**
 * Retry configuration for observer actions.
 */
export type RetryConfig = ObserverRetryConfig;

/**
 * Default retry configuration used when no retry is specified.
 */
export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  max_attempts: 3,
  backoff_strategy: "exponential",
  initial_delay_ms: 100,
  max_delay_ms: 60000,
};

/**
 * Configuration for the @Observer decorator.
 */
interface ObserverConfig {
  entity: string;
  event: string;
  actions: ObserverAction[];
  condition?: string;
  retry?: RetryConfig;
}

/**
 * Method decorator to register an observer with the schema registry.
 *
 * @param config - Observer configuration
 * @returns Method decorator
 */
export function Observer(config: ObserverConfig) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- legacy method decorator target
  return function (_target: any, propertyKey: string, _descriptor: PropertyDescriptor): void {
    SchemaRegistry.registerObserver(
      propertyKey,
      config.entity,
      config.event,
      config.actions,
      config.condition,
      config.retry
    );
  };
}

/**
 * Options for the webhook action factory.
 */
interface WebhookOptions {
  url_env?: string;
  headers?: Record<string, string>;
  body_template?: string;
}

/**
 * Create a webhook action.
 *
 * @param url - Static URL (mutually exclusive with url_env)
 * @param options - Additional options including url_env, headers, body_template
 * @returns Webhook action definition
 * @throws If neither url nor url_env is provided
 */
export function webhook(url?: string, options?: WebhookOptions): ObserverAction {
  if (url === undefined && options?.url_env === undefined) {
    throw new Error("Either url or url_env must be provided");
  }
  const action: ObserverAction & {
    headers: Record<string, string>;
    url?: string;
    url_env?: string;
    body_template?: string;
  } = {
    type: "webhook",
    headers: { "Content-Type": "application/json", ...(options?.headers ?? {}) },
  };
  if (url !== undefined) {
    action.url = url;
  }
  if (options?.url_env !== undefined) {
    action.url_env = options.url_env;
  }
  if (options?.body_template !== undefined) {
    action.body_template = options.body_template;
  }
  return action;
}

/**
 * Options for the slack action factory.
 */
interface SlackOptions {
  webhook_url?: string;
  webhook_url_env?: string;
}

/**
 * Create a Slack notification action.
 *
 * @param channel - Slack channel (e.g. "#orders")
 * @param message - Message template with {field} placeholders
 * @param options - Optional webhook URL or environment variable override
 * @returns Slack action definition
 */
export function slack(channel: string, message: string, options?: SlackOptions): ObserverAction {
  const action: ObserverAction & {
    channel: string;
    message: string;
    webhook_url_env: string;
    webhook_url?: string;
  } = {
    type: "slack",
    channel,
    message,
    webhook_url_env: options?.webhook_url_env ?? "SLACK_WEBHOOK_URL",
  };
  if (options?.webhook_url !== undefined) {
    action.webhook_url = options.webhook_url;
  }
  return action;
}

/**
 * Options for the email action factory.
 */
interface EmailOptions {
  from_email?: string;
}

/**
 * Create an email notification action.
 *
 * @param to - Recipient email address
 * @param subject - Email subject template with {field} placeholders
 * @param body - Email body template with {field} placeholders
 * @param options - Optional sender override
 * @returns Email action definition
 */
export function email(
  to: string,
  subject: string,
  body: string,
  options?: EmailOptions
): ObserverAction {
  const action: ObserverAction & {
    to: string;
    subject: string;
    body: string;
    from?: string;
  } = {
    type: "email",
    to,
    subject,
    body,
  };
  if (options?.from_email !== undefined) {
    action.from = options.from_email;
  }
  return action;
}
